.pre-commit-config.yaml

```diff
@@ -0,0 +1,19 @@
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.5.0
+    hooks:
+      - id: trailing-whitespace
+      - id: end-of-file-fixer
+      - id: check-yaml
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.1.14
+    hooks:
+      - id: ruff
+  - repo: https://github.com/PyCQA/isort
+    rev: 5.13.2
+    hooks:
+      - id: isort
+  - repo: https://github.com/psf/black
+    rev: 23.12.1
+    hooks:
+      - id: black
```

Comment:
I would not run black and ruff at the same time. The ruff config mimicks blak formatting, however there are some subtle differences.

Ultimately both formating and linting should be taken care of by ruff. Having both black and ruff doing the same thing will only lead to conflicts.

Since this PR should only affect new files the choice of formatter should not matter and as such I would vote for ruff being the sole formatting tool. Besides, the ``pyproject.toml`` already contains the formatting configuration for ruff.

Reply:
Does `ruff` perform the same formatting as `black`? Do they follow the same rules?

Reply:
Not necessarily. In principle it should be a ``black`` drop in replacement (see [docs](https://docs.astral.sh/ruff/formatter/#black-compatibility)). However, in a very limited number of cases it will fail the ``black`` checks. In an ideal word there would be a 'all in one' switch where the formatting provider is changed to ruff.

For reference, my IDE automatically applies ruff formatting and I never had issues. But I recall other people having issues.

In my opinion, ruff should be adopted as the default formatter moving forward and black should only be used case-by-case where the pipeline rejects the formatting.

Reply:
removed black